rust/wasmtime_bindings/src/test_memory_creator.rs

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+use anyhow::{anyhow, Result};
+use std::{cmp, ptr, slice, u32};
+use wasmtime::LinearMemory;
+
+const WASM_PAGE_SIZE: usize = 64 * 1024;
+
+extern "C" {
+    fn aligned_alloc(align: usize, size: usize) -> *mut u8;
+    fn free(ptr: *mut u8);
+}
+
+// The implementation below is the same as in ScyllaLinearMemory
+// and ScyllaMemoryCreator, with the addition of a conditional
+// failure of an allocation in grow_to().
+// It is used in testing instead of an alloc_failure_injector
+// because we only assert that the big allocations of WASM pages
+// can fail without aborting (as the allocations more likely to fail)
+// TODO: remove this code when avoiding crashes on allocation failures
+//       becomes possible in Rust
+pub struct TestScyllaLinearMemory {
+    ptr: *mut u8,
+    size: usize,
+    maximum_size: Option<usize>,
+    allocs: usize,
+    fail_after: usize,
+}
+
+// The entire TestScyllaLinearMemory is used only in a single thread,
+// because we're not sharing it between seastar shards
+unsafe impl Send for TestScyllaLinearMemory {}
+unsafe impl Sync for TestScyllaLinearMemory {}
+
+impl Drop for TestScyllaLinearMemory {
+    fn drop(&mut self) {
+        // previously allocated or reset to nullptr in grow_to()
+        unsafe { free(self.ptr) };
+    }
+}
+
+unsafe impl LinearMemory for TestScyllaLinearMemory {
```

Comment:
I managed to mostly fix this in the rebase.  The new behaviour of the `TestScyllaLinearMemory::grow_to` is now slightly different - the change isn't really important but I described it below anyway.
Because the duplication of the logic of deciding when we even need to perform any allocations has been removed, now, instead of inserting the failed alloc after we decided that it was needed using the mentioned logic, we now perform the alloc using the `ScyllaLinearMemory` and depending on the result we insert the failed alloc afterwards. 
A side effect of this is that the `TestScyllaLinearMemory::grow_to` will succeed even though it returns an `Err`, leaving the `TestScyllaLinearMemory` in an unexpected state, but we aren't planning to use it anyway later, so there's probably no harm in it.

Reply:
I'm not sure whether it's a problem either. The documentation of `LinearMemory` is not clear on what should happen with memory when grow_to fails ("Returns Err if memory can’t be grown by the specified amount of bytes").

If you are concerned about grow_to changing the memory pointer on failure, I guess you could split the current `ScyllaLinearMemory::grow_to` into two functions:

```rust
/// Validate and adjust the size to be allocated.
/// Returns None if realllocation is not needed
pub fn grow_prepare(&self, new_size: usize) -> Result<Option<usize>>;

/// Actually performs the reallocation.
/// Assumes that the size was returned from grow_prepare.
pub fn grow_perform(&mut self, new_size: usize) -> Result<()>;
```

Then you could use them to implement grow_to in both `ScyllaLinearMemory` and `TestScyllaLinearMemory`.

Reply:
I think I'll leave it as it is now. As you mentioned, LinearMemory is not clear on what should happen with memory when grow_to fails, and there is no reason for us to introduce a new rule for it - it would be even worse if we started depending on it and be surprised later.

If I were changing this, I would probably use your method with `grow_prepare` and `grow_perform`, however.